c_emulator/riscv_callbacks.cpp

```diff
@@ -0,0 +1,118 @@
+#include "riscv_callbacks.h"
+#include "riscv_config.h"
+#include <stdlib.h>
+#include <vector>
+
+std::vector<uint8_t> read_lbits(lbits val)
+{
+  std::vector<uint8_t> data(val.len);
+  mpz_export(data.data(), nullptr, -1, 1, 0, 0, *val.bits);
+  return data;
+}
+
+// Implementations of default callbacks for trace printing.
+//
+// The model assumes that these functions do not change the state of the model.
+unit mem_write_callback(uint64_t addr, uint64_t width, lbits value)
+{
+  if (config_print_mem_access) {
+    std::vector<uint8_t> lbits_data = read_lbits(value);
+    printf("mem[0x%.16lX] <- 0x", addr);
+    for (int i = width - 1; i >= 0; --i)
+      printf("%02hhX", lbits_data[i]);
+    printf("\n");
+  }
+#ifdef RVFI_DII
+  zrvfi_write(addr, width, value);
+#endif
+  return UNIT;
+}
+
+unit mem_read_callback(const char *type, uint64_t addr, uint64_t width,
+                       lbits value)
+{
+  if (config_print_mem_access) {
+    std::vector<uint8_t> lbits_data = read_lbits(value);
+    printf("mem[%s,0x%.16lX] -> 0x", type, addr);
```

Comment:
Ah I also noticed we probably want to use `fprintf(trace_log,` here and in (almost?) all the other cases.